### ISSUE TITLE
Pyright fails to resolve `type(None)` to the `None` singleton, preventing standard type narrowing

### ISSUE THREAD

**Issue by rob3c** [microsoft/pyright#11407]
### Describe the bug
When `type(None)` is assigned to a variable and used in a `Union` type hint, Pyright fails to resolve it to the `None` singleton. This prevents standard type narrowing (e.g., `if not obj.my_list`) from removing the `None` branch, causing elements to be inferred as `Unknown`.

### Code Sample

```python
import typing
none_type = type(None) # Alias for the None singleton

class MyClass:
    # Pyright treats this Union as list[str] | Unknown
    my_list: typing.Union[list[str], none_type]

def fn(o: MyClass):
    if not o.my_list:
        return
    
    # Expected: list[str]
    # Actual: list[str | Unknown]
    elements = [e for e in o.my_list]
```

### Environment
Python 3.13, Pyright/Pylance v2026.2.1

**Comment by gramster** [comment]
I'm confused; that `if` will throw an exception in the code you have provided. You have to actually initialize a `my_list` attribute, and then the type is correct.

**Comment by rob3c** [comment]
Sorry, I wrote this too quickly, and my minimal repro is too minimal lol!

Here's a real working example that still shows the designtime typing problem indicated above:

```python
import typing

none_type = type(None) # Alias for the `None` singleton

class MyClass:
    def __init__(self):
        # Pyright treats this Union as list[str] | Unknown
        self.my_list: typing.Union[list[str], none_type] = None

def fn(o: MyClass):
    if not o.my_list:
        return

    # Expected: list[str]
    # Actual: list[str | Unknown]
    elements = [e for e in o.my_list]


mc = MyClass()
fn(mc)
mc.my_list = ['a']
fn(mc)
```

i.e. Pylance/Pyright still breaks in the way I described, but the code actually works now.

**Comment by rob3c** [comment]
For added context where I ran into this problem (I use `None` directly in my own code without incident), it was using the [CVAT SDK](https://github.com/cvat-ai/cvat). Part of that library code is generated using templates for the [OpenAPI Generator](https://github.com/openapitools/openapi-generator).

Strictly speaking, the CVAT project is choosing to use the `none_type` indirection in their templates that Pylance/Pyright doesn't seem to be handling correctly. That said, the openapi-generator project itself seems to use it in their python generator, and CVAT likely just followed their example.

I've already opened a [issue](https://github.com/cvat-ai/cvat/issues/10531) in the CVAT repo as well FWIW. If you claim that Pylance/Pyright is working as intended, then it'll be especially important for them to update their templates. And I should probably create an issue with OpenAPI Generator also.

**Comment by rob3c** [comment]
Here's what **PEP 484** says about [Using `None`](https://peps.python.org/pep-0484/#using-none):
> When used in a type hint, the expression `None` is considered equivalent to `type(None)`.


Maybe CVAT's and OAPI Gen's use of `none_type` as above is correctly considered wrong by Pyright. But since PEP 484 says using `None` and `type(None)` in type hints are equivalent, why is this syntax ok:

```python
self.my_list = typing.Union[list[str], None] = None
```

but this one is not and causes Pyright to introduce the persistent `Unknown` type:
```Python
self.my_list: typing.Union[list[str], type(None)] = None
```

Also, previously in **PEP 483** in both the [fundamental building blocks](https://peps.python.org/pep-0483/#fundamental-building-blocks) and [Pragmatics](https://peps.python.org/pep-0483/#pragmatics) sections, the usage of `None` and `type(None)` is said to be equivalent, and they're specifically shown in `Union` type hints like this:

**fundamental building blocks:**
> **Optional[t1]**. Alias for `Union[t1, None]`, i.e. `Union[t1, type(None)]`.

**pragmatics:**
> Where a type is expected, `None` can be substituted for `type(None)`; e.g. `Union[t1, None]` == `Union[t1, type(None)]`.

**Comment by dchevell** [comment]
The problem isn't the use of `NoneType`, or conformance with the spec. The problem is you're assigning the result of a call expression to a variable, then trying to use that variable as a type hint. That's fundamentally invalid typing syntax. Consider:

```python
int_type = type(1)
value: int_type = 42
reveal_type(value)  # revealed type is "Unknown"
```
Same issue, nothing to do with `None`.

It's true that _"when used in a type hint, the expression None is considered equivalent to type(None)"_, but this is describing how the `None` type hint is evaluated by type checkers, it's not claiming this is a valid type hint. Static type hint evaluation doesn't run code, so you can't use normal variables, or call expressions, to create a type expression. 

The simplest answer is to not try to achieve this, and to file a bug / PR if you see someone else doing so. If you really want to alias it to a variable, you can create a Type Alias instead:

```python
# python 3.12+
type NoneType = None

# python 3.11 or below
from typing import TypeAlias
NoneType: TypeAlias = None
```

**Comment by rob3c** [comment]
@dchevell That's what I originally thought when I first saw that code in the third party libraries I mentioned I was quoting. However, that's not quite correct, which is why I chose this issue's subject the way I did.

[My comment quoting PEP 483 and PEP 484  just before your reply](https://github.com/microsoft/pyright/issues/11407#issuecomment-4357974729) most directly addresses the real issue mentioned in the subject:

Pyright is ok with this:
```python
import typing
ok: typing.Union[list[str], None] = None
```

But Pyright flags this with `reportInvalidTypeForm` even though it's directly quoted from PEP 484 and described as ok in PEP 483:
```python
import typing
broken: typing.Union[list[str], type(None)] = None
```

<img width="408" height="103" alt="Image" src="https://github.com/user-attachments/assets/7aca6277-c0cc-4ba7-a9b7-e92c97f5ec49" />

Maybe I should update the original issue, but as a reader of issues I prefer to have everything inline to see the full history and avoid confusion due to then out-of-context comments. However, I'm willing to update it if that aligns better with this repo's preferences.

In any case, the variable indirection was always questionable and masked the underlying issue: i.e. Pyright fails to resolve `type(None)` to the `None` singleton, preventing standard type narrowing.

I don't know python/pyright dev details in terms of what you consider spec conformance or not. My simplistic view is, if it's in a PEP it's in the spec, but maybe you can clarify that for me. If the PEP 483/484 usage I've quoted isn't valid for the python spec pyright works with, I'm happy to close the issue.

As an aside, I use `None` and the type alias syntax you mentioned directly in my own code and never run into these problems. Uncovering this bug in Pyright only happened when I ran into that odd usage in the third party libraries, so I'm reporting it here. The CVAT library devs are already fixing their usage, so I won't have this problem once it's merged.

**Comment by dchevell** [comment]
@rob3c I think you may be missing the point of my previous comment. `type(None)` is not valid syntax for a type annotation, no type checker would accept this. The PEP you're quoting doesn't say this is valid syntax, it says `None` is interpreted as equivalent to the value that `type(None)` would produce when run in python. Type checkers *don't run code*. The spec you're quoting doesn't mean what you think it does.

**Comment by rob3c** [comment]
@dchevell Thanks for the quick reply. I'm happy to here more details and understand those PEPs better, since I mentioned before what my simplistic interpretation of them is. However, "the spec you're quoting doesn't mean what you think it does" isn't very enlightening. Please elaborate, if you don't mind.

Here's a specific quote from the first bullet point in the  [PEP 483 pragmatics section](https://peps.python.org/pep-0483/#pragmatics) (my emphasis):
> Where a type is expected, None **can** be substituted for type(None); e.g. Union[t1, None] == Union[t1, type(None)]

By saying `None` **can** be substituted for `type(None)` where a type is expected, it sure looks like using `type(None)` is expected to be valid syntax. Using `None` directly instead seems like an alternative syntax that's also valid "where a type is expected".

I'm very interested in hearing exactly how this doesn't mean what I think it means. Thanks in advance!

**Comment by rob3c** [comment]
I'll go ahead and close this. It looks like this is the problem with only looking at the relevant PEPs to investigate the third party usage and not being privvy to insider discussions between PEP authors and actual language and type checker implementors as they hashed out the details outside the PEPs.

I did find some mypy issues that mention the misleading PEP language. Here's [the oldest one](https://github.com/python/mypy/issues/640) from 2015 that I saw before I stopped looking. The issue author (a mypy collaborator) says:
> Or maybe we should update PEP 484 to not suggest that type(None) is valid as a type, and None is the only correct spelling? There should one -- and preferably only one -- obvious way to do it etc.

to which another collaborator replies (before closing the issue):
> I don't think anyone cares

🤣 🤣 🤣 

I'm guessing disallowing the syntax is due to the technical issue of it being a callable, the PEP 483 usage was only meant to be suggestive as a placeholder, and fundamental limitations of early python 3 didn't have a good solution to handling the None type and other special cases yet as type hinting was developing.
